src-ghc/Pact/Types/ECC.hs

```diff
@@ -0,0 +1,214 @@
+
+
+
+
+--------- ECDSA HELPER FUNCTIONS ---------
+
+keccak256Hash :: ByteString -> ByteString
+keccak256Hash =
+  BS.pack . BA.unpack . (H.hash :: BA.Bytes -> H.Digest H.Keccak_256) . BA.pack . BS.unpack
+
+
+
+
+-- ECDSA Private Key must be 32 bytes and not begin with 0x00 (null byte)
+-- Assumes ByteString is not base 16.
+-- Source: https://kobl.one/blog/create-full-ethereum-keypair-and-address/
+
+importPrivate :: ByteString -> Maybe PrivateKey
+importPrivate bs | not startsNullByte &&
+                   BS.length bs == 32     = checkIfValid
+                 | otherwise              = Nothing
+  where startsNullByte =
+          (BS.take 1 bs) == (integerToBS 0x00)
+        i = bsToInteger bs
+        n = ECDSA.ecc_n (ECDSA.common_curve curveECDSA)
+        checkIfValid
+          | i >= 1 && i <= n   = Just $ PrivateKey curveECDSA i
+          | otherwise          = Nothing
+
+
+
+
+-- TODO: 65 bytes with 04 in front
+-- ECDSA Public Key must be uncompressed and 64 bytes long.
+-- Assumes ByteString is not base 16.
+-- Source: https://kobl.one/blog/create-full-ethereum-keypair-and-address/
+
+importPublic :: ByteString -> Maybe PublicKey
+importPublic bs | BS.length bs == 65 &&
+                  startsWithConstant4    = checkIfValid (BS.drop 1 bs)
+                | BS.length bs == 64     = checkIfValid bs
+                | otherwise              = Nothing
+  where  startsWithConstant4 =
+           (BS.take 1 bs) == (integerToBS 0x04)
+         point b = ECDSA.Point (bsToInteger xBS) (bsToInteger yBS)
+           where (xBS, yBS) = BS.splitAt 32 b
+         checkIfValid b
+           | isPointValid curveECDSA (point b) &&
+             not (isPointAtInfinity (point b))      = Just $ PublicKey curveECDSA (point b)
+           | otherwise                              = Nothing
+
+
+
+
+-- Assumes ByteString is not base 16.
+
+importSignature :: ByteString -> Maybe Signature
+importSignature bs | BS.length bs == 64   = Just makeSignature
+                   | otherwise            = Nothing
+  where (rBS, sBS) = BS.splitAt 32 bs
+        makeSignature = Signature (bsToInteger rBS) (bsToInteger sBS)
+
+
+
+
+integerToBS :: Integer -> ByteString
+integerToBS = i2osp
+
+
+bsToInteger :: ByteString -> Integer
+bsToInteger = os2ip
+
+
+
+
+{--
+integralToHexBS :: Integral a => a -> ByteString
+integralToHexBS a = BS.pack $ reverse $ go (quotRem a base) where
+  go (n,d) | n == 0 = [fromIntegral d]
+           | otherwise = (fromIntegral d):go (quotRem n base)
+  base = 256
+
+
+-- Generates an ECDSA nonce (`k`) deterministically according to RFC 6979.
+-- https://tools.ietf.org/html/rfc6979#section-3.2
+-- Based in this example:
+--   https://github.com/btcsuite/btcd/blob/master/btcec/signature.go#L455
+-- If k-signature returns zero should re-determine k
+deterministicNonce :: PrivateKey -> ByteString -> Int
+deterministicNonce = undefined
+--}
+
+
+
+
+--------- ECDSA TESTS ---------
```

Comment:
Should move these into unit tests if they are actually useful, or delete if they're not useful